diff -ru a/dlls/bcrypt/bcrypt_main.c b/dlls/bcrypt/bcrypt_main.c
--- a/dlls/bcrypt/bcrypt_main.c	2021-11-21 02:53:58.133096000 +0300
+++ b/dlls/bcrypt/bcrypt_main.c	2021-11-21 02:56:28.403257304 +0300
@@ -1703,6 +1703,7 @@
     case ALG_ID_ECDSA_P256:
         size = sizeof(BCRYPT_ECCKEY_BLOB) + 2 * 256 / 8;
         break;
+    case ALG_ID_ECDH_P384:
     case ALG_ID_ECDSA_P384:
         size = sizeof(BCRYPT_ECCKEY_BLOB) + 2 * 384 / 8;
         break;

