From a0c3ac854ac408247ec65ae0516fe9b791f9d4f9 Mon Sep 17 00:00:00 2001
From: Gijs Vermeulen <gijsvrm@gmail.com>
Date: Thu, 28 Jan 2021 16:30:18 +0100
Subject: [PATCH 1/2] helper_directx_dl: update link

---
 src/winetricks | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/winetricks b/src/winetricks
index b047dc6a..adbb7471 100755
--- a/src/winetricks
+++ b/src/winetricks
@@ -5755,7 +5755,8 @@ helper_directx_dl()
     # https://www.microsoft.com/en-us/download/details.aspx?id=9033
     # FIXME: none of the verbs that use this will show download status right
     # until file1 metadata is extended to handle common cache dir
-    w_download_to directx9 https://download.microsoft.com/download/E/E/1/EE17FF74-6C45-4575-9CF4-7FC2597ACD18/directx_feb2010_redist.exe f6d191e89a963d7cca34f169d30f49eab99c1ed3bb92da73ec43617caaa1e93f
+    # 2021/01/28: https://download.microsoft.com/download/E/E/1/EE17FF74-6C45-4575-9CF4-7FC2597ACD18/directx_feb2010_redist.exe
+    w_download_to directx9 https://fichiers.touslesdrivers.com/27356/directx_feb2010_redist.exe f6d191e89a963d7cca34f169d30f49eab99c1ed3bb92da73ec43617caaa1e93f

     DIRECTX_NAME=directx_feb2010_redist.exe
 }

From 0af3521e66d4b8b1d941710e7384a871cd801546 Mon Sep 17 00:00:00 2001
From: Gijs Vermeulen <gijsvrm@gmail.com>
Date: Thu, 28 Jan 2021 16:34:27 +0100
Subject: [PATCH 2/2] helper_directx_Jun2010: update link

---
 src/winetricks | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/winetricks b/src/winetricks
index adbb7471..4894fff8 100755
--- a/src/winetricks
+++ b/src/winetricks
@@ -5766,7 +5766,8 @@ helper_directx_Jun2010()
 {
     # June 2010 DirectX 9c User Redistributable
     # https://www.microsoft.com/en-us/download/details.aspx?id=8109
-    w_download_to directx9 https://download.microsoft.com/download/8/4/A/84A35BF1-DAFE-4AE8-82AF-AD2AE20B6B14/directx_Jun2010_redist.exe
+    # 2021/01/28: https://download.microsoft.com/download/8/4/A/84A35BF1-DAFE-4AE8-82AF-AD2AE20B6B14/directx_Jun2010_redist.exe
+    w_download_to directx9 https://fichiers.touslesdrivers.com/30604/directx_Jun2010_redist.exe 8c35296f330c0c4e431a7cf155013f3e4abec4a74efd68bb6bfcd76f08c46193

     DIRECTX_NAME=directx_Jun2010_redist.exe
 }
