--- wine-staging-2.19-orig/dlls/vulkan/vulkan_main.c	2017-10-27 11:41:21.372403970 +0200
+++ wine-staging-2.19-orig/dlls/vulkan/vulkan_main.c	2017-10-27 10:59:46.599503891 +0200
@@ -185,11 +185,11 @@ 
 
     TRACE( "(%p, %s)\n", instance, debugstr_a(pName) );
 
-    if (is_null_func( pName ))
-    {
-        FIXME( "%s not supported\n", debugstr_a(pName) );
-        return NULL;
-    }
+//    if (is_null_func( pName ))
+//    {
+//        FIXME( "%s not supported\n", debugstr_a(pName) );
+//        return NULL;
+//    }
 
     function = GetProcAddress( vulkan_module, pName );
     if (!function) FIXME( "missing function %s\n", debugstr_a(pName) );
--- wine-staging-2.19-orig/dlls/vulkan/vulkan_private.h	2017-10-27 11:41:21.379070619 +0200
+++ wine-staging-2.19-orig/dlls/vulkan/vulkan_private.h	2017-10-27 10:10:22.885354153 +0200
@@ -44,6 +44,7 @@ 
 #define VK_FALSE 0
 #define VK_STRUCTURE_TYPE_XLIB_SURFACE_CREATE_INFO_KHR 1000004000
 #define VK_STRUCTURE_TYPE_XCB_SURFACE_CREATE_INFO_KHR 1000005000
+#define VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2_KHR 1000059001
 
 #define VK_DESCRIPTOR_TYPE_SAMPLER                 0
 #define VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER  1
@@ -1865,6 +1866,13 @@ 
     VkPhysicalDeviceSparseProperties sparseProperties;
 } VkPhysicalDeviceProperties_host;
 
+typedef struct VkPhysicalDeviceProperties2KHR
+{
+    VkStructureType sType;
+    const void* pNext;
+    VkPhysicalDeviceProperties properties;
+} VkPhysicalDeviceProperties2KHR;
+
 typedef struct VkQueueFamilyProperties
 {
     VkQueueFlags queueFlags;
--- wine-staging-2.19-orig/dlls/vulkan/vulkan.spec	2017-10-27 11:41:21.365737321 +0200
+++ wine-staging-2.19-orig/dlls/vulkan/vulkan.spec	2017-10-27 10:35:48.624176879 +0200
@@ -134,6 +134,7 @@ 
 @ stdcall vkGetPhysicalDeviceImageFormatProperties(ptr long long long long long ptr)
 @ stdcall vkGetPhysicalDeviceMemoryProperties(ptr ptr)
 @ stdcall vkGetPhysicalDeviceProperties(ptr ptr)
+@ stdcall vkGetPhysicalDeviceProperties2KHR(ptr ptr)
 @ stdcall vkGetPhysicalDeviceQueueFamilyProperties(ptr ptr ptr)
 @ stdcall vkGetPhysicalDeviceSparseImageFormatProperties(ptr long long long long long ptr ptr)
 @ stdcall vkGetPhysicalDeviceSurfaceCapabilitiesKHR(ptr int64 ptr)
--- wine-staging-2.19-orig/dlls/vulkan/vulkan_thunks.c	2017-10-27 11:41:21.379070619 +0200
+++ wine-staging-2.19-orig/dlls/vulkan/vulkan_thunks.c	2017-10-27 10:39:57.990174328 +0200
@@ -3942,6 +3942,12 @@ 
     FIXME( "(%p, %p) not supported\n", physicalDevice, pProperties );
 }
 
+static void null_vkGetPhysicalDeviceProperties2KHR( VkPhysicalDevice physicalDevice,
+        VkPhysicalDeviceProperties2KHR *pProperties )
+{
+    FIXME( "(%p, %p) not supported\n", physicalDevice, pProperties );
+}
+
 static void null_vkGetPhysicalDeviceQueueFamilyProperties( VkPhysicalDevice physicalDevice,
         uint32_t *pQueueFamilyPropertyCount, VkQueueFamilyProperties *pQueueFamilyProperties )
 {
@@ -4422,6 +4428,8 @@ 
         VkPhysicalDeviceMemoryProperties_host * ) = null_vkGetPhysicalDeviceMemoryProperties;
 void (*p_vkGetPhysicalDeviceProperties)( VkPhysicalDevice, VkPhysicalDeviceProperties_host * ) =
         null_vkGetPhysicalDeviceProperties;
+void (*p_vkGetPhysicalDeviceProperties2KHR)( VkPhysicalDevice, VkPhysicalDeviceProperties2KHR * ) =
+        null_vkGetPhysicalDeviceProperties2KHR;
 void (*p_vkGetPhysicalDeviceQueueFamilyProperties)( VkPhysicalDevice, uint32_t *,
         VkQueueFamilyProperties * ) = null_vkGetPhysicalDeviceQueueFamilyProperties;
 void (*p_vkGetPhysicalDeviceSparseImageFormatProperties)( VkPhysicalDevice, VkFormat, VkImageType,
@@ -6598,6 +6606,18 @@ 
     return p_vkWaitForFences( device, fenceCount, pFences, waitAll, timeout );
 }
 
+
+void WINAPI vkGetPhysicalDeviceProperties2KHR( VkPhysicalDevice physicalDevice,
+        VkPhysicalDeviceProperties2KHR* pProperties )
+{
+    TRACE( "(%p, %p)\n", physicalDevice, pProperties );
+
+    pProperties->sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2_KHR;
+    pProperties->pNext = NULL;
+    vkGetPhysicalDeviceProperties( physicalDevice, &pProperties->properties );
+}
+
+
 struct function_entry
 {
     const char *name;
@@ -6746,6 +6766,7 @@ 
     DEFINE_FUNCTION( vkGetPhysicalDeviceImageFormatProperties )
     DEFINE_FUNCTION( vkGetPhysicalDeviceMemoryProperties )
     DEFINE_FUNCTION( vkGetPhysicalDeviceProperties )
+    DEFINE_FUNCTION( vkGetPhysicalDeviceProperties2KHR )
     DEFINE_FUNCTION( vkGetPhysicalDeviceQueueFamilyProperties )
     DEFINE_FUNCTION( vkGetPhysicalDeviceSparseImageFormatProperties )
     DEFINE_FUNCTION( vkGetPhysicalDeviceSurfaceCapabilitiesKHR )
--- a/dlls/ntdll/file.c	2017-11-09 18:37:38.433258842 +0100
+++ b/dlls/ntdll/file.c	2017-11-13 22:04:13.096568323 +0100
@@ -1188,8 +1188,9 @@ NTSTATUS WINAPI NtReadFileScatter
 
  error:
     if (needs_close) close( unix_handle );
-    if (status == STATUS_SUCCESS)
+    if (status == STATUS_SUCCESS || status == STATUS_END_OF_FILE)
     {
+        status = STATUS_PENDING;
         io_status->u.Status = status;
         io_status->Information = total;
         TRACE("= SUCCESS (%u)\n", total);

